lib/elixir/lib/calendar/duration.ex

```diff
@@ -0,0 +1,222 @@
+defmodule Duration do
+  @moduledoc """
+  The Duration type.
+  """
+
+  @default [year: 0, month: 0, week: 0, day: 0, hour: 0, minute: 0, second: 0, microsecond: 0]
+  @fields Keyword.keys(@default)
+
+  defstruct @default
+
+  @typedoc "Duration in calendar units"
+  @type t :: %Duration{
+          year: integer,
+          month: integer,
+          week: integer,
+          day: integer,
+          hour: integer,
+          minute: integer,
+          second: integer,
+          microsecond: integer
+        }
+
+  @typedoc "Individually valid Duration units"
+  @type unit ::
+          {:year, integer}
+          | {:month, integer}
+          | {:week, integer}
+          | {:day, integer}
+          | {:hour, integer}
+          | {:minute, integer}
+          | {:second, integer}
+          | {:microsecond, integer}
+
+  @doc """
+  Create `Duration` struct from valid duration units.
+
+  Returns `{:error, :invalid_duration}` when called with invalid units.
+
+  ## Examples
+
+      iex> Duration.new(month: 2)
+      {:ok, %Duration{month: 2}}
+      iex> Duration.new(months: 2)
+      {:error, :invalid_duration}
+
+  """
+  @spec new([unit]) :: {:ok, t} | {:error, :invalid_duration}
+  def new(units) do
+    case Keyword.validate(units, @fields) do
+      {:ok, units} ->
+        {:ok, struct(Duration, units)}
+
+      {:error, _invalid_keys} ->
+        {:error, :invalid_duration}
+    end
```

Comment:
I think we should raise on invalid options, because it is an error in the user input, not a semantic error. Similar to `Map.new(123)` or `Map.new([{:three, :element, :tuple])`.